<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>How to Extract Coordinates and Sequence from a FASTA File â€¢ rustytools</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="How to Extract Coordinates and Sequence from a FASTA File">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">rustytools</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.2</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Introduction</h6></li>
    <li><a class="dropdown-item" href="../articles/introduction.html">Getting Started with rustytools</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>FASTA file</h6></li>
    <li><a class="dropdown-item" href="../articles/getConsensus.html">How to Extract Coordinates and Sequence from a FASTA File</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Pairwise Alignment</h6></li>
    <li><a class="dropdown-item" href="../articles/seqAlign.html">How to Perform Sequence Alignment</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>PCHA</h6></li>
    <li><a class="dropdown-item" href="../articles/pcha.html">PCHA</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Magic</h6></li>
    <li><a class="dropdown-item" href="../articles/magic.html">Magic in Rust</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="http://twitter.com/scottfurlan" aria-label="Twitter"><span class="fa fa-twitter"></span></a></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/furlan-lab/rustytools" aria-label="GitHub"><span class="fa fa-github"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>How to Extract Coordinates and Sequence from a FASTA File</h1>
            
            <h4 data-toc-skip class="date">2024-06-04</h4>
      

      <div class="d-none name"><code>getConsensus.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<p>This vignette demonstrates how to extract contiguous consensus
sequence blocks from a FASTA file using <code>rustytools</code>, and how
to annotate them with genomic features from various sources such as
UCSC, GENCODE, and Ensembl.</p>
<p>This is particularly useful for processing FASTA files derived from
consensus calling pipelines (e.g., <code>samtools consensus</code>) and
for downstream variant or feature annotation.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"furlan-lab/rustytools"</span>, force <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<hr>
</div>
<div class="section level2">
<h2 id="load-required-packages">Load Required Packages<a class="anchor" aria-label="anchor" href="#load-required-packages"></a>
</h2>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/rm.html" class="external-link">rm</a></span><span class="op">(</span>list <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/Bimap-envirAPI.html" class="external-link">ls</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="op">{</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://furlan-lab.github.io/rustytools/">rustytools</a></span><span class="op">)</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/GenomicRanges" class="external-link">GenomicRanges</a></span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span></code></pre></div>
<hr>
</div>
<div class="section level2">
<h2 id="extracting-consensus-sequences">Extracting Consensus Sequences<a class="anchor" aria-label="anchor" href="#extracting-consensus-sequences"></a>
</h2>
<p>The <code><a href="../reference/get_consensus.html">get_consensus()</a></code> function extracts sequence blocks
containing A/C/T/G from a FASTA file and returns a <code>GRanges</code>
object with embedded sequence metadata.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">root</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/libPaths.html" class="external-link">.libPaths</a></span><span class="op">(</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="st">"rustytools/extdata"</span><span class="op">)</span></span>
<span><span class="va">tfasta</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">root</span>, <span class="st">"test.fa"</span><span class="op">)</span></span>
<span><span class="va">results</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_consensus.html">get_consensus</a></span><span class="op">(</span><span class="va">tfasta</span>, cores <span class="op">=</span> <span class="fl">1</span>, genome <span class="op">=</span> <span class="st">"toy"</span><span class="op">)</span></span>
<span><span class="va">results</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 7 ranges and 1 metadata column:</span></span>
<span><span class="co">##       seqnames    ranges strand | mcols.sequence</span></span>
<span><span class="co">##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; |    &lt;character&gt;</span></span>
<span><span class="co">##   [1]     chr1   238-251      * | ACTGACCTTAAGGT</span></span>
<span><span class="co">##   [2]    chr12   102-115      * | ACTGACCTTAAGGT</span></span>
<span><span class="co">##   [3]    chr12   595-608      * | ACTGACCTTAAGGT</span></span>
<span><span class="co">##   [4]     chr3   317-330      * | ACTGACCTTAAGGT</span></span>
<span><span class="co">##   [5]     chr4   168-181      * | ACTGACCTTAAGGT</span></span>
<span><span class="co">##   [6]     chr5   171-184      * | ACTGACCTTAAGGT</span></span>
<span><span class="co">##   [7]     chr6    97-110      * | ACTGACCTTAAGGT</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 7 sequences from toy genome</span></span></code></pre>
<p>The <code>cores</code> parameter defines how many threads to use
(typically one per chromosome). For toy examples, multi-core use is
unnecessary.</p>
<hr>
</div>
<div class="section level2">
<h2 id="running-on-large-fasta-files">Running on Large FASTA Files<a class="anchor" aria-label="anchor" href="#running-on-large-fasta-files"></a>
</h2>
<p>For real genomes, such as those from HLA typing or clinical
samples:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fasta</span> <span class="op">&lt;-</span> <span class="st">"~/Desktop/Junk/GENDX_HLA_Mix1_S1_L001_t1000.fa"</span></span></code></pre></div>
<p>To test on only the <strong>first few contigs</strong> before running
a full genome:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">results</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_consensus.html">get_consensus</a></span><span class="op">(</span><span class="va">fasta</span>, cores <span class="op">=</span> <span class="fl">16</span>, genome <span class="op">=</span> <span class="st">"hg38"</span>, test_with_n <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<hr>
</div>
<div class="section level2">
<h2 id="reducing-and-merging-consensus-ranges">Reducing and Merging Consensus Ranges<a class="anchor" aria-label="anchor" href="#reducing-and-merging-consensus-ranges"></a>
</h2>
<p>You can merge nearby regions using <code><a href="https://rdrr.io/pkg/IRanges/man/inter-range-methods.html" class="external-link">reduce()</a></code>:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">reduced_results</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/inter-range-methods.html" class="external-link">reduce</a></span><span class="op">(</span><span class="va">results</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, min.gapwidth <span class="op">=</span> <span class="fl">2L</span><span class="op">)</span></span></code></pre></div>
<hr>
</div>
<div class="section level2">
<h2 id="annotation-methods">Annotation Methods<a class="anchor" aria-label="anchor" href="#annotation-methods"></a>
</h2>
<div class="section level3">
<h3 id="ucsc-knowngene-txdb">1. UCSC KnownGene (TxDb)<a class="anchor" aria-label="anchor" href="#ucsc-knowngene-txdb"></a>
</h3>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">TxDb.Hsapiens.UCSC.hg38.knownGene</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">VariantAnnotation</span><span class="op">)</span></span>
<span></span>
<span><span class="va">txdb_hg38</span> <span class="op">&lt;-</span> <span class="va">TxDb.Hsapiens.UCSC.hg38.knownGene</span></span>
<span><span class="va">si</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqinfo</a></span><span class="op">(</span><span class="va">txdb_hg38</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Ensure correct seqinfo</span></span>
<span><span class="va">reduced_results</span><span class="op">@</span><span class="va">seqinfo</span> <span class="op">&lt;-</span> <span class="fu">GenomeInfoDb</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/Seqinfo-class.html" class="external-link">Seqinfo</a></span><span class="op">(</span></span>
<span>  seqnames <span class="op">=</span> <span class="va">si</span><span class="op">@</span><span class="va">seqnames</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">22</span><span class="op">]</span>,</span>
<span>  seqlengths <span class="op">=</span> <span class="va">si</span><span class="op">@</span><span class="va">seqlengths</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">22</span><span class="op">]</span>,</span>
<span>  isCircular <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="cn">NA</span>, <span class="fl">22</span><span class="op">)</span>,</span>
<span>  genome <span class="op">=</span> <span class="st">"hg38"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">annotated_results</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/VariantAnnotation/man/locateVariants-methods.html" class="external-link">locateVariants</a></span><span class="op">(</span><span class="va">reduced_results</span>, <span class="va">txdb_hg38</span>, <span class="fu"><a href="https://rdrr.io/pkg/VariantAnnotation/man/VariantType-class.html" class="external-link">AllVariants</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">annotated_results</span><span class="op">$</span><span class="va">LOCATION</span><span class="op">)</span></span>
<span><span class="va">annotated_results</span><span class="op">[</span><span class="va">annotated_results</span><span class="op">$</span><span class="va">LOCATION</span> <span class="op">==</span> <span class="st">"coding"</span><span class="op">]</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="gencode-gtf-annotation-with-txdbmaker">2. GENCODE GTF Annotation with <code>txdbmaker</code><a class="anchor" aria-label="anchor" href="#gencode-gtf-annotation-with-txdbmaker"></a>
</h3>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/txdbmaker" class="external-link">txdbmaker</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">gtf_file</span> <span class="op">&lt;-</span> <span class="st">"/Users/sfurlan/Library/CloudStorage/OneDrive-SharedLibraries-FredHutchinsonCancerCenter/Furlan_Lab - General/resources/refs/gencode.v46.annotation.gtf.gz"</span></span>
<span></span>
<span><span class="va">txdb</span> <span class="op">&lt;-</span> <span class="fu">makeTxDbFromGFF</span><span class="op">(</span><span class="va">gtf_file</span>, format <span class="op">=</span> <span class="st">"gtf"</span>, dbxrefTag <span class="op">=</span> <span class="st">"gene_name"</span><span class="op">)</span></span>
<span><span class="va">annotated_results2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/VariantAnnotation/man/locateVariants-methods.html" class="external-link">locateVariants</a></span><span class="op">(</span><span class="va">reduced_results</span>, <span class="va">txdb</span>, <span class="fu"><a href="https://rdrr.io/pkg/VariantAnnotation/man/VariantType-class.html" class="external-link">AllVariants</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">annotated_results2</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="ensembl-based-annotation-with-chippeakanno">3. Ensembl-based Annotation with <code>ChIPpeakAnno</code><a class="anchor" aria-label="anchor" href="#ensembl-based-annotation-with-chippeakanno"></a>
</h3>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">ChIPpeakAnno</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/jorainer/ensembldb" class="external-link">ensembldb</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">EnsDb.Hsapiens.v86</span><span class="op">)</span></span>
<span></span>
<span><span class="va">annoGR</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ChIPpeakAnno/man/toGRanges.html" class="external-link">toGRanges</a></span><span class="op">(</span><span class="va">EnsDb.Hsapiens.v86</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqlevelsStyle.html" class="external-link">seqlevelsStyle</a></span><span class="op">(</span><span class="va">reduced_results</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqlevelsStyle.html" class="external-link">seqlevelsStyle</a></span><span class="op">(</span><span class="va">annoGR</span><span class="op">)</span></span>
<span></span>
<span><span class="va">annotated_results3</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ChIPpeakAnno/man/annoPeaks.html" class="external-link">annoPeaks</a></span><span class="op">(</span><span class="va">reduced_results</span>, <span class="va">annoGR</span>, bindingType <span class="op">=</span> <span class="st">"fullRange"</span><span class="op">)</span></span>
<span><span class="va">annotated_results3</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/grep.html" class="external-link">grepl</a></span><span class="op">(</span><span class="st">"^HLA"</span>, <span class="va">annotated_results3</span><span class="op">$</span><span class="va">gene_name</span><span class="op">)</span><span class="op">]</span><span class="op">$</span><span class="va">gene_name</span></span></code></pre></div>
<hr>
</div>
</div>
<div class="section level2">
<h2 id="summary">Summary<a class="anchor" aria-label="anchor" href="#summary"></a>
</h2>
<ul>
<li>
<code><a href="../reference/get_consensus.html">get_consensus()</a></code> extracts contiguous sequence blocks
from FASTA files.</li>
<li>Results are returned as <code>GRanges</code> objects that are easily
annotated.</li>
<li>Annotations can be performed using UCSC TxDb, GENCODE GTFs, or
Ensembl databases.</li>
</ul>
<p>This workflow is ideal for consensus-based variant discovery, isoform
localization, or follow-up annotation after long-read assemblies.</p>
<hr>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="https://furlan-lab.github.io" class="external-link">Scott Furlan</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
