<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="rustytools">
<title>How to extract coordinates and sequence from a fasta file • rustytools</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="How to extract coordinates and sequence from a fasta file">
<meta property="og:description" content="rustytools">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">rustytools</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.9000</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <h6 class="dropdown-header" data-toc-skip>Get Consensus</h6>
    <a class="dropdown-item" href="../articles/getConsensus.html">How to extract coordinates and sequence from a fasta file</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="http://twitter.com/scottfurlan" aria-label="Twitter">
    <span class="fa fa-twitter"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/furlan-lab/rustytools" aria-label="GitHub">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>How to extract coordinates and sequence from a fasta file</h1>
            
            <h4 data-toc-skip class="date">2024-06-04</h4>
      
      
      <div class="d-none name"><code>getConsensus.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="installing-rust">Installing Rust<a class="anchor" aria-label="anchor" href="#installing-rust"></a>
</h2>
<p>First you need to have an updated Rust installation. Go to this <a href="https://www.rust-lang.org/tools/install" class="external-link">site</a> to learn how to
install Rust.</p>
</div>
<div class="section level2">
<h2 id="installing-rustytools">Installing rustytools<a class="anchor" aria-label="anchor" href="#installing-rustytools"></a>
</h2>
<p>You will need to have the devtools package installed…</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"scfurl/rustytools"</span>, force<span class="op">=</span><span class="cn">T</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="running-a-tool-to-extract-consensus-sequence-data">Running a tool to extract consensus sequence data<a class="anchor" aria-label="anchor" href="#running-a-tool-to-extract-consensus-sequence-data"></a>
</h2>
<p>Loading libraries</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/rm.html" class="external-link">rm</a></span><span class="op">(</span>list<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/ls.html" class="external-link">ls</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="op">{</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://scfurl.github.io/rustytools/" class="external-link">rustytools</a></span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="get-consensus-from-fasta">Get consensus from fasta<a class="anchor" aria-label="anchor" href="#get-consensus-from-fasta"></a>
</h2>
<p>This function will process a fasta file that has been processed using
a tool such as samtools consensus. The function will find all regions
(subsequences) of the fasta that have sequence (A, C, T, G) and output a
granges object containing the features and their sequences. It is
invoked as follows:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">root</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/libPaths.html" class="external-link">.libPaths</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">"rustytools/extdata"</span><span class="op">)</span></span>
<span><span class="va">tfasta</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">root</span>, <span class="st">"test.fa"</span><span class="op">)</span></span>
<span><span class="va">results</span><span class="op">&lt;-</span><span class="fu"><a href="../reference/get_consensus.html">get_consensus</a></span><span class="op">(</span><span class="va">tfasta</span>, cores<span class="op">=</span><span class="fl">1</span>, split<span class="op">=</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">results</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 7 ranges and 1 metadata column:</span></span>
<span><span class="co">##       seqnames    ranges strand | mcols.sequence</span></span>
<span><span class="co">##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; |    &lt;character&gt;</span></span>
<span><span class="co">##   [1]     chr1   238-251      * | ACTGACCTTAAGGT</span></span>
<span><span class="co">##   [2]    chr12   102-115      * | ACTGACCTTAAGGT</span></span>
<span><span class="co">##   [3]    chr12   595-608      * | ACTGACCTTAAGGT</span></span>
<span><span class="co">##   [4]     chr3   317-330      * | ACTGACCTTAAGGT</span></span>
<span><span class="co">##   [5]     chr4   168-181      * | ACTGACCTTAAGGT</span></span>
<span><span class="co">##   [6]     chr5   171-184      * | ACTGACCTTAAGGT</span></span>
<span><span class="co">##   [7]     chr6    97-110      * | ACTGACCTTAAGGT</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 7 sequences from hg38 genome</span></span></code></pre>
<p>In the above toy example we use the split argument to specify how
many chunks to break the sequence into for parallel processing. In this
case it doesn’t make sense to split the sequence into chunks because the
example only has about 500-600 bases for each contig. When running this
on real fasta sequence from a large genome, one would want to leverage
parallel processing. In this next block we show how to run get_consensus
on a big fasta. One can specify the name of the genome using the genome
argument. For hg38, we find that this takes about 1-3 minutes per
chromosome when using 16 cores on an intel iMac. We recommend splitting
the fasta contigs into 10000 chunks as below.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fasta</span><span class="op">=</span><span class="st">"abigfasta.fa"</span></span>
<span><span class="co">#fasta="~/Desktop/GENDX_HLA_Mix1_S1_L001_t1000.fa"</span></span>
<span><span class="va">results</span><span class="op">&lt;-</span><span class="fu"><a href="../reference/get_consensus.html">get_consensus</a></span><span class="op">(</span><span class="va">fasta</span>, cores<span class="op">=</span><span class="fl">16</span>, split<span class="op">=</span><span class="fl">10000</span>, genome<span class="op">=</span><span class="st">"hg38"</span><span class="op">)</span></span></code></pre></div>
<p>Before running a full fasta on all contigs, you can check the first n
contigs of a full sized fasta using the argument test_with_n. Note that
the full fasta will be read in to ensure appropriate seqinfo data</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fasta</span><span class="op">=</span><span class="st">"abigfasta.fa"</span></span>
<span><span class="co">#fasta="~/Desktop/GENDX_HLA_Mix1_S1_L001_t1000.fa"</span></span>
<span><span class="va">results</span><span class="op">&lt;-</span><span class="fu"><a href="../reference/get_consensus.html">get_consensus</a></span><span class="op">(</span><span class="va">fasta</span>, cores<span class="op">=</span><span class="fl">16</span>, split<span class="op">=</span><span class="fl">10000</span>, genome<span class="op">=</span><span class="st">"hg38"</span>, test_with_n <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="appendix">Appendix<a class="anchor" aria-label="anchor" href="#appendix"></a>
</h2>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.3.1 (2023-06-16)</span></span>
<span><span class="co">## Platform: x86_64-apple-darwin20 (64-bit)</span></span>
<span><span class="co">## Running under: macOS Ventura 13.6.7</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /Library/Frameworks/R.framework/Versions/4.3-x86_64/Resources/lib/libRblas.0.dylib </span></span>
<span><span class="co">## LAPACK: /Library/Frameworks/R.framework/Versions/4.3-x86_64/Resources/lib/libRlapack.dylib;  LAPACK version 3.11.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: America/Los_Angeles</span></span>
<span><span class="co">## tzcode source: internal</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">## [1] rustytools_0.0.0.9000</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##  [1] ade4_1.7-22             jsonlite_1.8.8          compiler_4.3.1         </span></span>
<span><span class="co">##  [4] Rcpp_1.0.12             seqinr_4.2-36           GenomicRanges_1.52.1   </span></span>
<span><span class="co">##  [7] bitops_1.0-7            parallel_4.3.1          jquerylib_0.1.4        </span></span>
<span><span class="co">## [10] systemfonts_1.0.6       IRanges_2.34.1          textshaping_0.3.7      </span></span>
<span><span class="co">## [13] yaml_2.3.8              fastmap_1.1.1           pbmcapply_1.5.1        </span></span>
<span><span class="co">## [16] R6_2.5.1                XVector_0.40.0          GenomeInfoDb_1.36.4    </span></span>
<span><span class="co">## [19] knitr_1.45              MASS_7.3-60.0.1         BiocGenerics_0.46.0    </span></span>
<span><span class="co">## [22] desc_1.4.3              GenomeInfoDbData_1.2.10 bslib_0.6.1            </span></span>
<span><span class="co">## [25] rlang_1.1.3             cachem_1.0.8            xfun_0.42              </span></span>
<span><span class="co">## [28] fs_1.6.3                sass_0.4.9              memoise_2.0.1          </span></span>
<span><span class="co">## [31] cli_3.6.2               pkgdown_2.0.7           magrittr_2.0.3         </span></span>
<span><span class="co">## [34] zlibbioc_1.46.0         digest_0.6.35           rstudioapi_0.15.0      </span></span>
<span><span class="co">## [37] lifecycle_1.0.4         S4Vectors_0.38.2        vctrs_0.6.5            </span></span>
<span><span class="co">## [40] evaluate_0.23           ragg_1.3.0              stats4_4.3.1           </span></span>
<span><span class="co">## [43] RCurl_1.98-1.14         rmarkdown_2.26          purrr_1.0.2            </span></span>
<span><span class="co">## [46] tools_4.3.1             htmltools_0.5.7</span></span></code></pre>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by <a href="https://furlan-lab.github.io" class="external-link">Scott Furlan</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
