<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Getting Started with rustytools • rustytools</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="Getting Started with rustytools">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">rustytools</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.2</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Introduction</h6></li>
    <li><a class="dropdown-item" href="../articles/introduction.html">Getting Started with rustytools</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>FASTA file</h6></li>
    <li><a class="dropdown-item" href="../articles/getConsensus.html">How to Extract Coordinates and Sequence from a FASTA File</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Pairwise Alignment</h6></li>
    <li><a class="dropdown-item" href="../articles/seqAlign.html">How to Perform Sequence Alignment</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>PCHA</h6></li>
    <li><a class="dropdown-item" href="../articles/pcha.html">PCHA</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Magic</h6></li>
    <li><a class="dropdown-item" href="../articles/magic.html">Magic in R and Rust</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="http://twitter.com/scottfurlan" aria-label="Twitter"><span class="fa fa-twitter"></span></a></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/furlan-lab/rustytools" aria-label="GitHub"><span class="fa fa-github"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Getting Started with rustytools</h1>
            
      

      <div class="d-none name"><code>introduction.Rmd</code></div>
    </div>

    
    
<!-- Load MathJax for inline math support -->
<script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml.js"></script><div class="section level2">
<h2 id="installing-rust">Installing Rust<a class="anchor" aria-label="anchor" href="#installing-rust"></a>
</h2>
<p>First you need to have an updated Rust installation. Go to this <a href="https://www.rust-lang.org/tools/install" class="external-link">site</a> to learn how to
install Rust.</p>
</div>
<div class="section level2">
<h2 id="installing-rustytools">Installing rustytools<a class="anchor" aria-label="anchor" href="#installing-rustytools"></a>
</h2>
<p>You will need to have the devtools package installed…</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"furlan-lab/rustytools"</span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="compilation-errors-with-openblas">Compilation errors with openblas<a class="anchor" aria-label="anchor" href="#compilation-errors-with-openblas"></a>
</h3>
<p>When building R packages that depend on BLAS/LAPACK functionality
(e.g., via <strong>openblas-src</strong> or <strong>RcppEigen</strong>),
you may encounter linker errors such as:</p>
<pre><code>ld: -lto_library library filename must be 'libLTO.dylib'
OpenBLAS build failed: Subprocess returns with non-zero status: 2</code></pre>
<p>or</p>
<pre><code> thread 'main' panicked at /private/var/folders/63/z3dzfmg53g31179qxhwt5cq00000gn/T/RtmpIhMjfg/R.INSTALL17d5e6ffc8bf0/rustytools/src/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/openblas-src-0.10.11/build.rs:218:13:
  OpenBLAS build failed: Subprocess returns with non-zero status: 2</code></pre>
<p>These errors typically arise because <strong>openblas-src</strong> is
attempting to compile OpenBLAS from source using a GCC-fortran toolchain
that injects a <code>-lto_library</code> flag incompatible with the
Apple linker. However, the <strong>openblas-src</strong> crate in Rust
or package builds in R defaults to building its own copy of OpenBLAS,
invoking <code>gcc</code>/<code>gfortran</code> and ultimately passing
<code>-lto_library</code> to <code>ld</code>.</p>
<p>Because Apple’s <code>ld</code> expects
<code>-lto_library libLTO.dylib</code> (not
<code>-lto_library /usr/local/...</code>), the link step fails.</p>
<p>Alternatively you may have non-standard installation sites of
openblas, such as those on high-performance clusters.</p>
<p>At the moment, we have found only one foolproof method of
compilation.</p>
<p>First, make sure you have <a href="http://www.openmathlib.org/OpenBLAS/" class="external-link">openblas</a> installed on
your system. Second, clone rustytools in a shell in a directory you
would like to install the source file i.e.:</p>
<pre><code>cd yourdir/
git clone https://github.com/furlan-lab/rustytools.git
</code></pre>
<p>Then you have a few options:</p>
<div class="section level5">
<h5 id="easiest---declare-system-variable-then-install-repo">Easiest - declare system variable, then install repo<a class="anchor" aria-label="anchor" href="#easiest---declare-system-variable-then-install-repo"></a>
</h5>
<pre><code>export OPENBLAS_DIR=/your_openblas_dir

export OPENBLAS_DIR=/opt/homebrew/opt/openblas
# On Apple Silicon this is typically: /opt/homebrew/opt/openblas
# On Intel Macs this is typically: /usr/local/opt/openblas
# On an HPC this might be something like: /app/software/OpenBLAS/0.3.27-GCC-13.3.0</code></pre>
<p>Then compile rustytools</p>
<pre><code>cd rustytools
R CMD install .</code></pre>
</div>
<div class="section level4">
<h4 id="project-level--renviron">Project-level <code>.Renviron</code><a class="anchor" aria-label="anchor" href="#project-level--renviron"></a>
</h4>
<p>Edit the file named <code>.Renviron</code> in your package root:</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a><span class="co"># On Apple Silicon:</span></span>
<span id="cb7-2"><a href="#cb7-2" tabindex="-1"></a><span class="bu">export</span> <span class="va">OPENBLAS_DIR</span><span class="op">=</span>/opt/homebrew/opt/openblas</span>
<span id="cb7-3"><a href="#cb7-3" tabindex="-1"></a><span class="co"># On Intel Macs:</span></span>
<span id="cb7-4"><a href="#cb7-4" tabindex="-1"></a><span class="bu">export</span> <span class="va">OPENBLAS_DIR</span><span class="op">=</span>/usr/local/opt/openblas</span></code></pre></div>
<div class="section level5">
<h5 id="user-level--renviron">User-level <code>.Renviron</code><a class="anchor" aria-label="anchor" href="#user-level--renviron"></a>
</h5>
<p>Create/edit <code>~/.Renviron</code> so that the variables are always
available, even in non-interactive sessions (e.g.,
<code>R CMD INSTALL</code>, RStudio Build pane):</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a><span class="bu">export</span> <span class="va">OPENBLAS_DIR</span><span class="op">=</span>/opt/homebrew/opt/openblas</span>
<span id="cb8-2"><a href="#cb8-2" tabindex="-1"></a><span class="co"># On Intel Macs:</span></span>
<span id="cb8-3"><a href="#cb8-3" tabindex="-1"></a><span class="bu">export</span> <span class="va">OPENBLAS_DIR</span><span class="op">=</span>/usr/local/opt/openblas</span></code></pre></div>
</div>
<div class="section level5">
<h5 id="changing-error-message">Changing error message<a class="anchor" aria-label="anchor" href="#changing-error-message"></a>
</h5>
<p>If the nature of the error message changes to something like that
seen below, delete the target directory in rustytools repo and try
again</p>
<pre><code>        cargo build --lib --release --manifest-path=./rust/Cargo.toml --target-dir ./rust/target
   Compiling openblas-src v0.10.11
   Compiling lax v0.15.0
error: could not find native static library `openblas`, perhaps an -L flag is missing?

error: could not compile `openblas-src` (lib) due to 1 previous error</code></pre>
</div>
</div>
</div>
</div>
<div class="section level2">
<h2 id="overview-of-sequence-alignment">Overview of Sequence Alignment<a class="anchor" aria-label="anchor" href="#overview-of-sequence-alignment"></a>
</h2>
<p>This vignette introduces the <code><a href="../reference/align.html">align()</a></code> function from the
<code>rustytools</code> package, which provides a Rust-backed,
high-performance implementation of the Smith-Waterman algorithm. This
algorithm supports global, local, semiglobal, and fully custom alignment
modes using affine gap penalties.</p>
<p>The underlying engine is a generalized variant of the Smith-Waterman
algorithm provided by the <code>bio-edit</code> crate, which offers:</p>
<ul>
<li>Match/mismatch scoring via custom or predefined functions</li>
<li>Affine gap penalties (gap open + gap extension)</li>
<li>Flexible boundary behavior via customizable clipping penalties</li>
</ul>
</div>
<div class="section level2">
<h2 id="alignment-modes">Alignment Modes<a class="anchor" aria-label="anchor" href="#alignment-modes"></a>
</h2>
<p><code>rustytools</code> supports three classic modes and a flexible
custom mode:</p>
<ul>
<li>
<strong>Global alignment</strong>: No clipping is allowed; aligns
full length of both sequences.</li>
<li>
<strong>Local alignment</strong>: Finds the highest-scoring
subsequences.</li>
<li>
<strong>Semiglobal alignment</strong>: Global on one sequence, local
on the other (e.g., full query to partial reference).</li>
<li>
<strong>Custom alignment</strong>: Full control over boundary
conditions via clip penalties.</li>
</ul>
</div>
<div class="section level2">
<h2 id="example-semiglobal-alignment">Example: Semiglobal Alignment<a class="anchor" aria-label="anchor" href="#example-semiglobal-alignment"></a>
</h2>
<p>Here is an example aligning a query sequence to a longer
reference:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://furlan-lab.github.io/rustytools/">rustytools</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">query</span> <span class="op">&lt;-</span> <span class="st">"ACCGTGGAT"</span></span>
<span><span class="va">reference</span> <span class="op">&lt;-</span> <span class="st">"AAAAACCGTTGAT"</span></span>
<span></span>
<span><span class="co"># Perform semiglobal alignment</span></span>
<span><span class="va">score</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/align.html">align</a></span><span class="op">(</span><span class="va">query</span>, <span class="va">reference</span>, atype <span class="op">=</span> <span class="st">"semi-global"</span>, verbose <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span>
<span><span class="va">score</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 7</span></span></code></pre>
<p>This returns an alignment score that considers:</p>
<ul>
<li>Matches = +1</li>
<li>Mismatches = -1</li>
<li>Gap open = -5</li>
<li>Gap extend = -1</li>
<li>No penalty for clipping the start/end of the reference</li>
</ul>
</div>
<div class="section level2">
<h2 id="custom-alignment-configuration">Custom Alignment Configuration<a class="anchor" aria-label="anchor" href="#custom-alignment-configuration"></a>
</h2>
<p>Custom alignments allow full specification of prefix/suffix
penalties. Example:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">score</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/align.html">align</a></span><span class="op">(</span></span>
<span>  <span class="va">query</span>,</span>
<span>  <span class="va">reference</span>,</span>
<span>  atype <span class="op">=</span> <span class="st">"custom"</span>,</span>
<span>  match_score <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  mismatch_score <span class="op">=</span> <span class="op">-</span><span class="fl">3</span>,</span>
<span>  gap_open <span class="op">=</span> <span class="op">-</span><span class="fl">5</span>,</span>
<span>  gap_extend <span class="op">=</span> <span class="op">-</span><span class="fl">1</span>,</span>
<span>  xclip_prefix <span class="op">=</span> <span class="op">-</span><span class="fl">10</span>, <span class="co"># allow prefix skips in query</span></span>
<span>  xclip_suffix <span class="op">=</span> <span class="op">-</span><span class="fl">9999</span>, <span class="co"># require alignment to end</span></span>
<span>  yclip_prefix <span class="op">=</span> <span class="fl">0</span>,     <span class="co"># allow local match to ref</span></span>
<span>  yclip_suffix <span class="op">=</span> <span class="fl">0</span></span>
<span><span class="op">)</span></span>
<span><span class="va">score</span></span></code></pre></div>
<p>This setting mimics semiglobal alignment where the query can skip
leading bases but must align to the end.</p>
</div>
<div class="section level2">
<h2 id="use-cases-in-genomics">Use Cases in Genomics<a class="anchor" aria-label="anchor" href="#use-cases-in-genomics"></a>
</h2>
<p>Common applications include:</p>
<ul>
<li>Mapping short variable sequences (e.g. CDR3s) to consensus
references</li>
<li>Comparing VDJ segments to known clone sequences</li>
<li>Imputing mutations into reference haplotypes</li>
</ul>
<p>See the <a href="articles/seqAlign.html">sequence alignment
vignette</a> for an example using single cell RNA seq data in B cell
leukemia</p>
</div>
<div class="section level2">
<h2 id="summary">Summary<a class="anchor" aria-label="anchor" href="#summary"></a>
</h2>
<p>The alignment engine in <code>rustytools</code> is highly flexible
and efficient. It is particularly well-suited to bioinformatics tasks
involving fuzzy or partial sequence matching. For large datasets or many
pairwise alignments, this Rust backend provides a dramatic speed
advantage over traditional R-native solutions.</p>
</div>
<div class="section level2">
<h2 id="other-tools">Other Tools<a class="anchor" aria-label="anchor" href="#other-tools"></a>
</h2>
<p><code>rustytools</code> includes additional fast backends for:</p>
<ul>
<li>
<strong>FASTA reading</strong> – random access to large genome files
See the <a href="getConsensus.html">Fasta vignette</a>
</li>
<li>
<strong>MAGIC</strong> – imputation of scRNA-seq using diffusion
maps See the <a href="magic.html">MAGIC vignette</a>
</li>
<li>
<strong>PCHA</strong> – archetypal decomposition of matrices See the
<a href="pcha.html">PCHA vignette</a>
</li>
</ul>
<p>Each tool is designed for high performance, especially on large-scale
single-cell or genome-wide data.</p>
</div>
<div class="section level2">
<h2 id="support-and-citation">Support and Citation<a class="anchor" aria-label="anchor" href="#support-and-citation"></a>
</h2>
<p>Developed by the <a href="https://furlan-lab.github.io/" class="external-link">Furlan
Lab</a> at Fred Hutchinson Cancer Center.</p>
<p>For questions or issues, please file a GitHub issue or contact the
maintainers directly.</p>
<hr>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="https://furlan-lab.github.io" class="external-link">Scott Furlan</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
