<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Magic in Rust • rustytools</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="Magic in Rust">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">rustytools</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.2</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Pairwise Alignment</h6></li>
    <li><a class="dropdown-item" href="../articles/seqAlign.html">How to perform sequence alignment</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>PCHA</h6></li>
    <li><a class="dropdown-item" href="../articles/pcha.html">PCHA</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Magic</h6></li>
    <li><a class="dropdown-item" href="../articles/magic.html">Magic in Rust</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="http://twitter.com/scottfurlan" aria-label="Twitter"><span class="fa fa-twitter"></span></a></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/furlan-lab/rustytools" aria-label="GitHub"><span class="fa fa-github"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Magic in Rust</h1>
            
            <h4 data-toc-skip class="date">2025-05-06</h4>
      

      <div class="d-none name"><code>magic.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="installing-rust">Installing Rust<a class="anchor" aria-label="anchor" href="#installing-rust"></a>
</h2>
<p>First you need to have an updated Rust installation. Go to this <a href="https://www.rust-lang.org/tools/install" class="external-link">site</a> to learn how to
install Rust.</p>
</div>
<div class="section level2">
<h2 id="installing-rustytools">Installing rustytools<a class="anchor" aria-label="anchor" href="#installing-rustytools"></a>
</h2>
<p>You will need to have the devtools package installed…</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"furlan-lab/rustytools"</span>, force<span class="op">=</span><span class="cn">T</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="overview-of-magic-implemented-in-rust">Overview of Magic implemented in Rust<a class="anchor" aria-label="anchor" href="#overview-of-magic-implemented-in-rust"></a>
</h2>
<p><em>(“Markov Affinity-based Graph Imputation of Cells” – van Dijk et
al., 2018)</em></p>
<table class="table">
<colgroup>
<col width="12%">
<col width="7%">
<col width="80%">
</colgroup>
<thead><tr class="header">
<th>Symbol</th>
<th>Size</th>
<th>Meaning</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><strong>X</strong></td>
<td><em>n × g</em></td>
<td>raw (or PCA-reduced) expression; rows = cells, cols = genes</td>
</tr>
<tr class="even">
<td><strong>d(i,j)</strong></td>
<td>–</td>
<td>Euclidean (or cosine) distance between cell <em>i</em> and
<em>j</em> in PCA space</td>
</tr>
<tr class="odd">
<td><strong>σᵢ</strong></td>
<td>–</td>
<td>
<strong>local scale</strong> for cell <em>i</em> (distance to its
<em>k</em>-th nearest neighbour)</td>
</tr>
<tr class="even">
<td><strong>A</strong></td>
<td><em>n × n</em></td>
<td>symmetrised, locally scaled affinity (“heat”) matrix</td>
</tr>
<tr class="odd">
<td><strong>D</strong></td>
<td><em>n × n</em></td>
<td>diagonal degree matrix, <strong>D</strong><sub>ii</sub> =
∑<sub>j</sub><strong>A</strong><sub>ij</sub>
</td>
</tr>
<tr class="even">
<td><strong>P = D⁻¹A</strong></td>
<td><em>n × n</em></td>
<td>row-stochastic Markov transition matrix</td>
</tr>
<tr class="odd">
<td><strong>t</strong></td>
<td>–</td>
<td>diffusion time (walk length)</td>
</tr>
<tr class="even">
<td><strong>Ŷ</strong></td>
<td><em>n × g</em></td>
<td>imputed expression after diffusion</td>
</tr>
</tbody>
</table>
<hr>
<div class="section level4">
<h4 id="k-nn-graph-in-latent-space">1 k-NN graph in latent space<a class="anchor" aria-label="anchor" href="#k-nn-graph-in-latent-space"></a>
</h4>
<p>Compute <em>k</em> nearest neighbours in the <em>m</em>-dimensional
PCA space (typically <em>k</em> ≈ 30, <em>m</em> ≈ 100).</p>
</div>
<div class="section level4">
<h4 id="adaptive-gaussian-kernel">2 Adaptive Gaussian kernel<a class="anchor" aria-label="anchor" href="#adaptive-gaussian-kernel"></a>
</h4>
<p>For each edge (<em>i</em>, <em>j</em>) in that graph set</p>
<p><span class="math display">$$
A_{ij}
\;=\;
\exp\!\Bigl[ -\tfrac{d(i,j)^2}{\sigma_i\,\sigma_j}\Bigr]
\quad\text{and}\quad
A_{ij}=0\;\text{if \(j\) is not a kNN of \(i\)} .
$$</span></p>
<p>Local bandwidths σᵢ make the kernel <strong>anisotropic</strong> so
dense and sparse regions of the manifold are treated equally.</p>
<div class="section level5">
<h5 id="symmetrise">Symmetrise<a class="anchor" aria-label="anchor" href="#symmetrise"></a>
</h5>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>A</mi><mspace width="0.278em"></mspace><mo>←</mo><mspace width="0.278em"></mspace><mfrac displaystyle="false"><mn>1</mn><mn>2</mn></mfrac><mspace width="0.167em"></mspace><mrow><mo stretchy="true" form="prefix">(</mo><mi>A</mi><mo>+</mo><msup><mi>A</mi><mrow><mspace width="-0.167em"></mspace><mi>⊤</mi></mrow></msup><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">
A \;←\; \tfrac12\,(A + A^{\!\top})
</annotation></semantics></math></p>
</div>
</div>
<div class="section level4">
<h4 id="markov-normalisation">3 Markov normalisation<a class="anchor" aria-label="anchor" href="#markov-normalisation"></a>
</h4>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>P</mi><mspace width="0.278em"></mspace><mo>=</mo><mspace width="0.278em"></mspace><msup><mi>D</mi><mrow><mi>−</mi><mn>1</mn></mrow></msup><mi>A</mi><mo>,</mo><mspace width="1.0em"></mspace><msub><mi>D</mi><mrow><mi>i</mi><mi>i</mi></mrow></msub><mo>=</mo><munder><mo>∑</mo><mi>j</mi></munder><msub><mi>A</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><mi>.</mi></mrow><annotation encoding="application/x-tex">
P \;=\; D^{-1}A,
\quad
D_{ii} = \sum_{j} A_{ij}.
</annotation></semantics></math></p>
<p>Now each row of <strong>P</strong> sums to 1 → one step of a random
walk on the cell graph.</p>
</div>
<div class="section level4">
<h4 id="diffusion-raise-to-power-t">4 Diffusion (raise to power <em>t</em>)<a class="anchor" aria-label="anchor" href="#diffusion-raise-to-power-t"></a>
</h4>
<p>There are two equivalent views:</p>
<ul>
<li>
<p><strong>Spectral</strong> Diagonalise <strong>P</strong> = Ψ Λ
Ψ⁻¹ with eigenvalues 1 = λ₀ &gt; λ₁ ≥ λ₂ … Then</p>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msup><mi>P</mi><mi>t</mi></msup><mspace width="0.278em"></mspace><mo>=</mo><mspace width="0.278em"></mspace><mi>Ψ</mi><mspace width="0.167em"></mspace><msup><mi>Λ</mi><mi>t</mi></msup><mspace width="0.167em"></mspace><msup><mi>Ψ</mi><mrow><mi>−</mi><mn>1</mn></mrow></msup><mo>,</mo><mspace width="2.0em"></mspace><msup><mi>Λ</mi><mi>t</mi></msup><mo>=</mo><mo>diag</mo><mrow><mo stretchy="true" form="prefix">(</mo><msubsup><mi>λ</mi><mi>k</mi><mi>t</mi></msubsup><mo stretchy="true" form="postfix">)</mo></mrow><mi>.</mi></mrow><annotation encoding="application/x-tex">
P^{t} \;=\; \Psi\,\Lambda^{t}\,\Psi^{-1},
\qquad
\Lambda^{t} = \operatorname{diag}(\lambda_k^{t}).
</annotation></semantics></math></p>
<p>Small eigenmodes (high-frequency noise) decay as λₖᵗ.</p>
</li>
<li><p><strong>Random-walk</strong> Entry (i,j) of <strong>Pᵗ</strong>
is the probability that a random walk of length <em>t</em> starting at
cell <em>i</em> ends in cell <em>j</em>.</p></li>
</ul>
</div>
<div class="section level4">
<h4 id="imputation-by-heat-propagation">5 Imputation by heat propagation<a class="anchor" aria-label="anchor" href="#imputation-by-heat-propagation"></a>
</h4>
<p>Apply the diffusion operator to every gene vector:</p>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><menclose notation="box"><mrow><mspace width="0.278em"></mspace><mover><mi>X</mi><mo accent="true">̂</mo></mover><mo>=</mo><msup><mi>P</mi><mi>t</mi></msup><mspace width="0.167em"></mspace><mi>X</mi><mspace width="0.278em"></mspace></mrow></menclose><mspace width="1.0em"></mspace><mo>⇒</mo><mspace width="1.0em"></mspace><msub><mover><mi>x</mi><mo accent="true">̂</mo></mover><mrow><mi>i</mi><mi>g</mi></mrow></msub><mo>=</mo><munderover><mo>∑</mo><mrow><mi>j</mi><mo>=</mo><mn>1</mn></mrow><mi>n</mi></munderover><msub><mrow><mo stretchy="true" form="prefix">(</mo><msup><mi>P</mi><mi>t</mi></msup><mo stretchy="true" form="postfix">)</mo></mrow><mrow><mi>i</mi><mi>j</mi></mrow></msub><mspace width="0.278em"></mspace><msub><mi>x</mi><mrow><mi>j</mi><mi>g</mi></mrow></msub><mi>.</mi></mrow><annotation encoding="application/x-tex">
\boxed{\;
\hat X = P^{t}\,X
\;}
\quad\Longrightarrow\quad
\hat x_{ig} = \sum_{j=1}^{n} (P^{t})_{ij}\; x_{jg}.
</annotation></semantics></math></p>
<p>Each imputed expression value becomes a weighted average over the
<em>t</em>-step neighbourhood, smoothing drop-outs while respecting
manifold structure.</p>
<hr>
</div>
<div class="section level3">
<h3 id="relation-to-the-continuous-heat-equation">Relation to the continuous heat equation<a class="anchor" aria-label="anchor" href="#relation-to-the-continuous-heat-equation"></a>
</h3>
<p>On a smooth data manifold <strong>M</strong>, the generator
<strong>L</strong> = <em>I</em> – <strong>P</strong> approximates the
Laplace–Beltrami operator Δ<sub>M</sub>. Diffusion time <em>t</em>
therefore controls how far one solves the heat-equation</p>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mfrac><mrow><mi>∂</mi><mi>u</mi></mrow><mrow><mi>∂</mi><mi>t</mi></mrow></mfrac><mspace width="0.278em"></mspace><mo>=</mo><mspace width="0.278em"></mspace><mo>−</mo><msub><mi>Δ</mi><mi>M</mi></msub><mspace width="0.167em"></mspace><mi>u</mi><mo>,</mo><mspace width="2.0em"></mspace><mi>u</mi><mrow><mo stretchy="true" form="prefix">(</mo><mn>0</mn><mo stretchy="true" form="postfix">)</mo></mrow><mo>=</mo><mi>X</mi><mi>.</mi></mrow><annotation encoding="application/x-tex">
\frac{\partial u}{\partial t} \;=\; -\Delta_M\,u, \qquad u(0)=X.
</annotation></semantics></math></p>
<p>MAGIC performs this on a <em>graph</em> and stops at a user-chosen
<em>t</em> (often 2–6); too small → little denoising, too large →
over-smoothing.</p>
<hr>
</div>
<div class="section level3">
<h3 id="where-our-rust-kernel-fits">Where our Rust kernel fits<a class="anchor" aria-label="anchor" href="#where-our-rust-kernel-fits"></a>
</h3>
<p>Steps 1–3 (k-NN, kernel, Markov row-normalisation) happen in
R/Python. The Rust function <code>diffuse_expr_r()</code> you now have
implements <strong>step 4 + 5</strong>:</p>
<pre class="text"><code>(P, X, t)  ──►  Ŷ = Pᵗ · X</code></pre>
<p>It expects <strong>P</strong> already row-stochastic; raising to
power <em>t</em> and the matrix–vector multiply are done in parallel for
speed and memory efficiency.</p>
</div>
</div>
<div class="section level2">
<h2 id="how-to-run-magic-in-rrust">How to run MAGIC in R/Rust<a class="anchor" aria-label="anchor" href="#how-to-run-magic-in-rrust"></a>
</h2>
<p>First we take some data, subset to the tumor cells that express
modest levels of CD33. Lets see how they look after Magic.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="op">{</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/samuel-marsh/scCustomize" class="external-link">scCustomize</a></span><span class="op">)</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://furlan-lab.github.io/rustytools/">rustytools</a></span><span class="op">)</span>   <span class="co"># your package exposing diffuse_expr_r()</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://magrittr.tidyverse.org" class="external-link">magrittr</a></span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="va">seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"~/Fred Hutch Cancer Center/Furlan_Lab - General/experiments/patient_marrows/DB1/cds/221022_DB1_seu_5p_with_cell_labels_SNF.RDS"</span><span class="op">)</span></span>
<span><span class="va">seu</span><span class="op">$</span><span class="va">sb</span> <span class="op">&lt;-</span> <span class="va">seu</span><span class="op">$</span><span class="va">geno</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="st">"0"</span> <span class="op">&amp;</span> <span class="va">seu</span><span class="op">$</span><span class="va">seurat_clusters</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"0"</span>, <span class="st">"1"</span>, <span class="st">"2"</span>, <span class="st">"3"</span>, <span class="st">"4"</span>, <span class="st">"11"</span><span class="op">)</span></span>
<span><span class="co">#DimPlot(seu, group.by = "sb")</span></span>
<span></span>
<span><span class="va">seu</span> <span class="op">&lt;-</span> <span class="va">seu</span><span class="op">[</span>,<span class="va">seu</span><span class="op">$</span><span class="va">sb</span><span class="op">]</span></span>
<span><span class="va">seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/NormalizeData.html" class="external-link">NormalizeData</a></span><span class="op">(</span><span class="va">seu</span>, verbose <span class="op">=</span> <span class="cn">F</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://satijalab.org/seurat/reference/ScaleData.html" class="external-link">ScaleData</a></span><span class="op">(</span>verbose <span class="op">=</span> <span class="cn">F</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindVariableFeatures.html" class="external-link">FindVariableFeatures</a></span><span class="op">(</span>verbose <span class="op">=</span> <span class="cn">F</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunPCA.html" class="external-link">RunPCA</a></span><span class="op">(</span>npcs <span class="op">=</span> <span class="fl">100</span>, verbose <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/ElbowPlot.html" class="external-link">ElbowPlot</a></span><span class="op">(</span><span class="va">seu</span>, ndims <span class="op">=</span> <span class="fl">100</span><span class="op">)</span></span></code></pre></div>
<p><img src="magic_files/figure-html/unnamed-chunk-2-1.png" width="100%"></p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindNeighbors.html" class="external-link">FindNeighbors</a></span><span class="op">(</span><span class="va">seu</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">35</span>, verbose <span class="op">=</span> <span class="cn">F</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindClusters.html" class="external-link">FindClusters</a></span><span class="op">(</span>verbose <span class="op">=</span> <span class="cn">F</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunUMAP.html" class="external-link">RunUMAP</a></span><span class="op">(</span>dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">35</span>, n.epochs <span class="op">=</span> <span class="fl">500</span>, verbose <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">seu</span><span class="op">)</span></span></code></pre></div>
<p><img src="magic_files/figure-html/unnamed-chunk-2-2.png" width="100%"></p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/seurat_magic.html">seurat_magic</a></span><span class="op">(</span><span class="va">seu</span>, alpha <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://satijalab.github.io/seurat-object/reference/DefaultAssay.html" class="external-link">DefaultAssay</a></span><span class="op">(</span><span class="va">seu</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"RNA"</span></span>
<span><span class="fu"><a href="https://samuel-marsh.github.io/scCustomize/reference/FeaturePlot_scCustom.html" class="external-link">FeaturePlot_scCustom</a></span><span class="op">(</span><span class="va">seu</span>, <span class="st">"CD33"</span><span class="op">)</span></span></code></pre></div>
<p><img src="magic_files/figure-html/unnamed-chunk-2-3.png" width="100%"></p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.github.io/seurat-object/reference/DefaultAssay.html" class="external-link">DefaultAssay</a></span><span class="op">(</span><span class="va">seu</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"MAGIC"</span></span>
<span><span class="fu"><a href="https://samuel-marsh.github.io/scCustomize/reference/FeaturePlot_scCustom.html" class="external-link">FeaturePlot_scCustom</a></span><span class="op">(</span><span class="va">seu</span>, <span class="st">"CD33"</span><span class="op">)</span></span></code></pre></div>
<p><img src="magic_files/figure-html/unnamed-chunk-2-4.png" width="100%"></p>
</div>
<div class="section level2">
<h2 id="changing-parameters">Changing parameters<a class="anchor" aria-label="anchor" href="#changing-parameters"></a>
</h2>
<p>An alpha of 1 might be too much. We can dial it back with an alpha of
0.3 and even further by dropping t to 2.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/seurat_magic.html">seurat_magic</a></span><span class="op">(</span><span class="va">seu</span>, alpha <span class="op">=</span> <span class="fl">0.3</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://satijalab.github.io/seurat-object/reference/DefaultAssay.html" class="external-link">DefaultAssay</a></span><span class="op">(</span><span class="va">seu</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"MAGIC"</span></span>
<span><span class="fu"><a href="https://samuel-marsh.github.io/scCustomize/reference/FeaturePlot_scCustom.html" class="external-link">FeaturePlot_scCustom</a></span><span class="op">(</span><span class="va">seu</span>, <span class="st">"CD33"</span><span class="op">)</span></span></code></pre></div>
<p><img src="magic_files/figure-html/unnamed-chunk-3-1.png" width="100%"></p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/seurat_magic.html">seurat_magic</a></span><span class="op">(</span><span class="va">seu</span>, alpha <span class="op">=</span> <span class="fl">0.3</span>, t <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://satijalab.github.io/seurat-object/reference/DefaultAssay.html" class="external-link">DefaultAssay</a></span><span class="op">(</span><span class="va">seu</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"MAGIC"</span></span>
<span><span class="fu"><a href="https://samuel-marsh.github.io/scCustomize/reference/FeaturePlot_scCustom.html" class="external-link">FeaturePlot_scCustom</a></span><span class="op">(</span><span class="va">seu</span>, <span class="st">"CD33"</span><span class="op">)</span></span></code></pre></div>
<p><img src="magic_files/figure-html/unnamed-chunk-3-2.png" width="100%"></p>
</div>
<div class="section level2">
<h2 id="appendix">Appendix<a class="anchor" aria-label="anchor" href="#appendix"></a>
</h2>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.4.3 (2025-02-28)</span></span>
<span><span class="co">## Platform: aarch64-apple-darwin20</span></span>
<span><span class="co">## Running under: macOS Sequoia 15.4.1</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRblas.0.dylib </span></span>
<span><span class="co">## LAPACK: /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: America/Los_Angeles</span></span>
<span><span class="co">## tzcode source: internal</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">## [1] future_1.40.0      magrittr_2.0.3     rustytools_0.0.2   scCustomize_3.0.1 </span></span>
<span><span class="co">## [5] Seurat_5.2.1       SeuratObject_5.0.2 sp_2.2-0          </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] RColorBrewer_1.1-3      shape_1.4.6.1           rstudioapi_0.17.1      </span></span>
<span><span class="co">##   [4] jsonlite_2.0.0          spatstat.utils_3.1-3    ggbeeswarm_0.7.2       </span></span>
<span><span class="co">##   [7] farver_2.1.2            rmarkdown_2.29          zlibbioc_1.52.0        </span></span>
<span><span class="co">##  [10] GlobalOptions_0.1.2     fs_1.6.5                ragg_1.4.0             </span></span>
<span><span class="co">##  [13] vctrs_0.6.5             ROCR_1.0-11             spatstat.explore_3.4-2 </span></span>
<span><span class="co">##  [16] paletteer_1.6.0         janitor_2.2.1           forcats_1.0.0          </span></span>
<span><span class="co">##  [19] htmltools_0.5.8.1       sass_0.4.9              sctransform_0.4.1      </span></span>
<span><span class="co">##  [22] parallelly_1.43.0       KernSmooth_2.23-26      bslib_0.9.0            </span></span>
<span><span class="co">##  [25] htmlwidgets_1.6.4       desc_1.4.3              ica_1.0-3              </span></span>
<span><span class="co">##  [28] plyr_1.8.9              lubridate_1.9.4         plotly_4.10.4          </span></span>
<span><span class="co">##  [31] zoo_1.8-14              cachem_1.1.0            igraph_2.1.4           </span></span>
<span><span class="co">##  [34] mime_0.13               lifecycle_1.0.4         pkgconfig_2.0.3        </span></span>
<span><span class="co">##  [37] Matrix_1.7-3            R6_2.6.1                fastmap_1.2.0          </span></span>
<span><span class="co">##  [40] GenomeInfoDbData_1.2.13 snakecase_0.11.1        fitdistrplus_1.2-2     </span></span>
<span><span class="co">##  [43] shiny_1.10.0            digest_0.6.37           colorspace_2.1-1       </span></span>
<span><span class="co">##  [46] rematch2_2.1.2          S4Vectors_0.44.0        patchwork_1.3.0        </span></span>
<span><span class="co">##  [49] tensor_1.5              RSpectra_0.16-2         irlba_2.3.5.1          </span></span>
<span><span class="co">##  [52] GenomicRanges_1.58.0    textshaping_1.0.0       labeling_0.4.3         </span></span>
<span><span class="co">##  [55] progressr_0.15.1        timechange_0.3.0        spatstat.sparse_3.1-0  </span></span>
<span><span class="co">##  [58] httr_1.4.7              polyclip_1.10-7         abind_1.4-8            </span></span>
<span><span class="co">##  [61] compiler_4.4.3          withr_3.0.2             fastDummies_1.7.5      </span></span>
<span><span class="co">##  [64] MASS_7.3-65             tools_4.4.3             vipor_0.4.7            </span></span>
<span><span class="co">##  [67] lmtest_0.9-40           beeswarm_0.4.0          httpuv_1.6.15          </span></span>
<span><span class="co">##  [70] future.apply_1.11.3     goftest_1.2-3           glue_1.8.0             </span></span>
<span><span class="co">##  [73] nlme_3.1-168            promises_1.3.2          grid_4.4.3             </span></span>
<span><span class="co">##  [76] Rtsne_0.17              ade4_1.7-23             cluster_2.1.8.1        </span></span>
<span><span class="co">##  [79] reshape2_1.4.4          generics_0.1.3          seqinr_4.2-36          </span></span>
<span><span class="co">##  [82] gtable_0.3.6            spatstat.data_3.1-6     tidyr_1.3.1            </span></span>
<span><span class="co">##  [85] data.table_1.17.0       XVector_0.46.0          BiocGenerics_0.52.0    </span></span>
<span><span class="co">##  [88] spatstat.geom_3.3-6     RcppAnnoy_0.0.22        ggrepel_0.9.6          </span></span>
<span><span class="co">##  [91] RANN_2.6.2              pillar_1.10.2           stringr_1.5.1          </span></span>
<span><span class="co">##  [94] ggprism_1.0.5           spam_2.11-1             RcppHNSW_0.6.0         </span></span>
<span><span class="co">##  [97] later_1.4.2             circlize_0.4.16         splines_4.4.3          </span></span>
<span><span class="co">## [100] dplyr_1.1.4             lattice_0.22-7          survival_3.8-3         </span></span>
<span><span class="co">## [103] deldir_2.0-4            tidyselect_1.2.1        miniUI_0.1.1.1         </span></span>
<span><span class="co">## [106] pbapply_1.7-2           knitr_1.50              gridExtra_2.3          </span></span>
<span><span class="co">## [109] IRanges_2.40.1          scattermore_1.2         stats4_4.4.3           </span></span>
<span><span class="co">## [112] xfun_0.52               matrixStats_1.5.0       UCSC.utils_1.2.0       </span></span>
<span><span class="co">## [115] stringi_1.8.7           lazyeval_0.2.2          yaml_2.3.10            </span></span>
<span><span class="co">## [118] evaluate_1.0.3          codetools_0.2-20        tibble_3.2.1           </span></span>
<span><span class="co">## [121] cli_3.6.4               uwot_0.2.3              pbmcapply_1.5.1        </span></span>
<span><span class="co">## [124] xtable_1.8-4            reticulate_1.42.0       systemfonts_1.2.2      </span></span>
<span><span class="co">## [127] munsell_0.5.1           jquerylib_0.1.4         GenomeInfoDb_1.42.3    </span></span>
<span><span class="co">## [130] Rcpp_1.0.14             globals_0.16.3          spatstat.random_3.3-3  </span></span>
<span><span class="co">## [133] png_0.1-8               ggrastr_1.0.2           spatstat.univar_3.1-2  </span></span>
<span><span class="co">## [136] parallel_4.4.3          pkgdown_2.1.1           ggplot2_3.5.2          </span></span>
<span><span class="co">## [139] dotCall64_1.2           listenv_0.9.1           viridisLite_0.4.2      </span></span>
<span><span class="co">## [142] scales_1.3.0            ggridges_0.5.6          purrr_1.0.4            </span></span>
<span><span class="co">## [145] rlang_1.1.6             cowplot_1.1.3</span></span></code></pre>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="https://furlan-lab.github.io" class="external-link">Scott Furlan</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
